hack/install-protobuf.sh

```diff
@@ -0,0 +1,58 @@
+#!/bin/sh
+set -e
+
+usage() {
+  this=$1
+  cat <<EOF
+$this: download and install protobuf
+
+Usage: $this [-b <bindir>] [<version>]
+  -b sets bindir or installation directory, defaults to ${HOME}/bin
+
+EOF
+  exit 2
+}
+
+parse_args() {
+  # BINDIR is ${HOME}/bin unless set by ENV
+  # overridden by flag below
+
+  BINDIR=${BINDIR:-${HOME}/bin}
+  while getopts "b:h?x" arg; do
+    case "$arg" in
+      b) BINDIR="$OPTARG" ;;
+      h | \?) usage "$0" ;;
+      x) set -x ;;
+    esac
+  done
+  shift $((OPTIND - 1))
+  VERSION=$1
+  if [ "$VERSION" = "" ]; then
+      usage "$0"
+  fi
+  # if version starts with 'v', remove it
+  VERSION=${VERSION#v}
+}
+
+source $(dirname $0)/library.sh
+
+PB_REL="https://github.com/protocolbuffers/protobuf/releases"
+OS=$(uname_os)
+ARCH=$(uname_arch)
+
+parse_args "$@"
+
+echo "OS: $OS  ARCH: $ARCH"
+
+tmp=$(mktemp -d)
+trap 'rm -rf ${tmp}' EXIT
+
+BINARY_URL=$PB_REL/download/v${VERSION}/protoc-${VERSION}-${OS}-${ARCH}.zip
+if  [ "$OS" = "darwin" ]; then
```

Comment:
@KeranYang - can you run `make codegen` on your M1 to see if it works?

@afugazzotto - can you also run it on your Windows to see what will happen?

Reply:
Documenting my steps to make it work:

1. Ran `make codegen`

Got the following error:

```
make codegen
./hack/generate-proto.sh
++++++++++++++++++++++++++++++++
++++ GENERATING PROTO FILES ++++
++++++++++++++++++++++++++++++++
OS: darwin  ARCH: arm64
Downloading https://github.com/protocolbuffers/protobuf/releases/download/v27.2/protoc-27.2-osx-universal_binary.zip
Archive:  /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.ovNQwhmJFE/protoc-27.2-darwin-arm64.zip
   creating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/bin/
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/bin/protoc  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/readme.txt  
   creating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/
   creating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/
   creating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/source_context.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/descriptor.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/java_features.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/timestamp.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/empty.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/duration.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/field_mask.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/cpp_features.proto  
   creating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/compiler/
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/compiler/plugin.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/api.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/struct.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/type.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/any.proto  
  inflating: /var/folders/hn/0grvmkss0wxdh1xdvyzwbdl00000gq/T/tmp.3T50Zi13cr/include/google/protobuf/wrappers.proto  
pkg/apis/proto/daemon/daemon.proto:22:1: warning: Import google/protobuf/empty.proto is unused.
./hack/update-codegen.sh
+++++++++++++++++++++++++
++++ RUNNING CODEGEN ++++
+++++++++++++++++++++++++
go: github.com/numaproj/numaflow/pkg/apis/proto/daemon imports
	github.com/grpc-ecosystem/grpc-gateway/runtime: missing go.sum entry for module providing package github.com/grpc-ecosystem/grpc-gateway/runtime (imported by github.com/numaproj/numaflow/pkg/apis/proto/daemon); to add:
	go get github.com/numaproj/numaflow/pkg/apis/proto/daemon
go: github.com/numaproj/numaflow/pkg/apis/proto/daemon imports
	github.com/grpc-ecosystem/grpc-gateway/utilities: missing go.sum entry for module providing package github.com/grpc-ecosystem/grpc-gateway/utilities (imported by github.com/numaproj/numaflow/pkg/apis/proto/daemon); to add:
	go get github.com/numaproj/numaflow/pkg/apis/proto/daemon
make: *** [codegen] Error 1
```

2. Ran `go get *` to update the dependency and Ran `make codegen` again, got 

```
+++++++++++++++++++++++++++
++++ UPDATING API DOCS ++++
+++++++++++++++++++++++++++
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!! Please upgrade pandoc to version 3.2.1 with - brew upgrade pandoc !!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
make: *** [codegen] Error 1
```

3. Ran `brew upgrade pandoc` and Ran `make codegen` again

It works.

The generated changes are as follow:

```
git status
On branch xxx
Your branch is up to date with 'origin/xxx'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   go.mod
	modified:   go.sum
	modified:   pkg/apis/proto/daemon/daemon.pb.go
	modified:   pkg/apis/proto/daemon/daemon.pb.gw.go
	modified:   pkg/apis/proto/daemon/daemon_grpc.pb.go
	modified:   pkg/apis/proto/isb/message.pb.go
	modified:   pkg/apis/proto/wmb/wmb.pb.go
```

Please let me know if I should check in these changes.

Reply:
Running `make codegen` completed with no issues for me.

Reply:
Synced offline, the root cause is that my local protoc-gen has an old version and the script is implemented in a way that if protoc-gen presents, it will skip installing. The solution could be to always install the latest.